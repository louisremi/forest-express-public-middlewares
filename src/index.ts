import { Request, Response } from "express"

export function addTimezone(timezone = process.env.TZ) {
  return (request: Request, response: Response, next: () => any) => {
    // This query param is required and client cannot alter it
    if (timezone) {
      request.query.timezone = timezone
    }

    next()
  }
}

type Whitelist = {
  [model: string]: string[]
}

type Fields = {
  [model: string]: string
}

export function whitelistFields(
  whitelist: Whitelist,
  warn: boolean | Function = true,
) {
  return (request: Request, response: Response, next: () => any) => {
    if (!request.query.fields) {
      request.query.fields = Object.keys(whitelist).reduce(
        (fields, modelName) => {
          fields[modelName] = whitelist[modelName].join(",")

          return fields
        },
        {} as Fields,
      )

      return next()
    }

    Object.keys(request.query.fields).forEach((modelName) => {
      if (!whitelist[modelName]) {
        delete (request.query.fields as Fields)[modelName]

        warn &&
          (typeof warn === "function" ? warn : console.warn)(
            `model \`${modelName}\` won't be included in the response as it isn't whitelisted`,
          )

        return
      }

      ;(request.query.fields as Fields)[modelName] = (request.query
        .fields as Fields)[modelName]
        .split(",")
        .filter((fieldName) => {
          if (whitelist[modelName].includes(fieldName)) {
            return true
          }

          warn &&
            (typeof warn === "function" ? warn : console.warn)(
              `field \`${modelName}.${fieldName}\` won't be included in the response as it isn't whitelisted`,
            )
        })
        .join(",")
    })

    next()
  }
}
