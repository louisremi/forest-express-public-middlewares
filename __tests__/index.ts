import { addTimezoneMiddleware, whitelistFields } from "../src/index"

describe("addTimezoneMiddleware", () => {
  it("constantly overwrites the query timezone", () => {
    const expectedTz = "Europe/Paris"
    const middleware = addTimezoneMiddleware(expectedTz)
    const req: any = { query: { timezone: "US/Pacific" } }

    middleware(req, {} as any, () => {})

    expect(req.query.timezone).toBe(expectedTz)
  })
})

describe("whitelistFields", () => {
  it("removes non-whitelisted fields from query params and executes warning function for each field removed", () => {
    const whitelist = {
      user: ["fullName", "age", "gender"],
      room: ["number", "name", "size"],
    }
    const warnings: string[] = []
    const warn = (msg: string) => warnings.push(msg)
    const middleware = whitelistFields(whitelist, warn)
    const req: any = {
      query: {
        fields: {
          user: "fullName,fullAddress",
          room: "number,name,rent",
          city: "name,postalCode",
        },
      },
    }

    middleware(req, {} as any, () => {})

    expect(req.query.fields).toEqual({
      user: "fullName",
      room: "number,name",
    })
    expect(warnings).toMatchSnapshot()
  })

  it("turns the whitelist into `fields` query params when absent from the query", () => {
    const whitelist = {
      user: ["fullName", "age", "gender"],
      room: ["number", "name", "size"],
    }
    const middleware = whitelistFields(whitelist)
    const req: any = { query: {} }

    middleware(req, {} as any, () => {})

    expect(req.query).toEqual({
      fields: {
        user: "fullName,age,gender",
        room: "number,name,size",
      },
    })
  })
})
