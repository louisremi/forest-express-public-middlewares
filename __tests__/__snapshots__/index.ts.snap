// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`whitelistFields removes non-whitelisted fields from query params and executes warning function for each field removed 1`] = `
Array [
  "field \`user.fullAddress\` won't be included in the response as it isn't whitelisted",
  "field \`room.rent\` won't be included in the response as it isn't whitelisted",
  "model \`city\` won't be included in the response as it isn't whitelisted",
]
`;
